proposals/p2173.md

```diff
@@ -0,0 +1,616 @@
+    introduced associated types, with the intent that a mechanism to constrain
+    their value would be provided in a later proposal.
+-   [#818: generics details 3](https://github.com/carbon-language/carbon-lang/pull/818)
+    introduced `where` constraints for generics, with an open question for
+    whether and how we might restrict constraints in order to give a decidable
+    type equality rule.
+    -   #818 represents the status quo of approved proposals. It has both `=`
+        and `==` where constraints, which are similar to, but somewhat different
+        from, those in this proposal.
+    -   `=` constraints require a concrete type on their right-hand side. The
+        left-hand side is not restricted. The type-of-type of the left-hand side
+        is changed to match the right-hand side.
+    -   `==` constraints do not affect the type-of-type of either operand.
+    -   As an ansewr to the question of how to produce a decidable type equality
+        rule, both kinds of constraints are applied automatically, but only at a
+        depth of one constraint. Neither form is transitive but both can be
+        transitively extended using `observe` declarations.
+-   [#2070: always `==` not `=` in `where` clauses](https://github.com/carbon-language/carbon-lang/pull/2070)
+    describes an attempt to unify the syntax and semantics of `=` and `==`
+    constraints, under which we would always merge the type-of-types of the
+    operands. That proposal has not yet been accepted, and this proposal intends
+    to supersede it.
+
+## Proposal
+
+Replace the existing `where A == B` constraint with two different kinds of
+constraint:
+
+-   `where .A = T` specifies a _rewrite_ constraint. If the associated constant
+    `.A` is named as a member of a type with this constraint, it is rewritten to
+    `T`. This means that the interface of the type changes, and the behavior is
+    in all respects as if `T` had been specified directly.
+-   `where X == Y` specifies a _same-type_ constraint. This is treated much like
+    `where X is SameAs(Y)`. Here, `SameAs` is a hypothetical constraint that is
+    reflexive (`T is SameAs(T)` for all `T`), commutative (`T is SameAs(U)` if
+    and only if `U is SameAs(T)`), and not implementable by the developer.
+    Moreover, if `F` is any pure type function, `T is SameAs(U)` implies
+    `F(T) is SameAs(F(U)`). `observe` statements can be used to form transitive
+    `SameAs` relations. Same-type constraints are not used automatically by the
+    language rules for any purpose, but there is an `ImplicitAs(T)` impl for
+    types implementing `SameAs(T)`.
+
+A type implements `C where .A = T` if and only if it implements
+`C where .A == T`, assuming both constraints are valid – the inhabitants of
+these two type-of-types are the same, but they are different type-of-types and
+provide different interfaces for corresponding type values.
+
+## Details
```

Comment:
Are these points anywhere in the proposal now?

- The "use same rewrite rule twice" error is triggered by invalid type expressions, not invalid constraint definitions. It would be ideal if the "rewrite ends up applying to itself after other rewrites have been applied" rule was sufficient, so errors were always diagnosed as early as possible, but this may not be possible while maintaining efficiency.
- Defining the interaction between `=` and `is` constraints. It isn't clear whether and when `is` constraints are rewritten by `=` rewrites. (Does it matter the order of the `is` constraint and the `=` rewrite? Making them mutually exclusive doesn't seem viable. I think the answer is that `is` constraints are still enforced as requirements but don't affect name lookup, except maybe possibly when they apply to the type of the ultimate destination of rewrites.)
- I think this proposal should explicitly state how this is an experiment to see when `==` is needed, and recommend using `=` over `==` where possible.

Reply:
I think these points are all now addressed.